### PR TITLE
Error modal changes

### PR DESCRIPTION
**Pull request by amitjha21**
This PR address #1256. 
More description can be found at Issue under Summary.
Corresponsing `ermrestjs` PR is ermrestjs#579.

This takes care of adding errors and changing error modal as per the table displayname.
  
Some of the error scenarios

|   |   |
| ------------- | ------------- |
| No record found  | https://dev.isrd.isi.edu/~amit/chaise/record/#1/isa:dataset/id=2693456  |
| Conflict  |  https://dev.isrd.isi.edu/~amit/chaise/record/#1/isa:dataset/id1=2693456  |
| invalid filter | https://dev.isrd.isi.edu/~amit/chaise/record/#1/isa:dataset/id |
| Deletion conflict | https://dev.isrd.isi.edu/~amit/chaise/record/#1/isa:dataset/id=269 |
| Deletion with inline table | https://dev.isrd.isi.edu/~amit/chaise/record/#39473/product-record:accommodation/id=2002 |  
| Conflict in recordedit| https://dev.isrd.isi.edu/~amit/chaise/recordedit/#39473/product-record:accommodation/id=2002 |

### CODE REVIEW

**Comment by amitjha21** [comment]
Working on fixing these travis errors. `testerrors` passes locally but have some URL issue on Travis as it has a different structure on Travis environment.

**Comment by amitjha21** [comment]
This is working locally but fails at Travis.

**Comment by amitjha21** [comment]
All builds are working now.

File: common/errors.js [hunk 1]
```diff
@@ -140,11 +143,12 @@
     }])
 
     // Factory for each error type
-    .factory('ErrorService', ['AlertsService', 'errorNames', 'Session', '$log', '$rootScope', '$uibModal', '$window', 'errorMessages', 'Errors',
-          function ErrorService(AlertsService, errorNames, Session, $log, $rootScope, $uibModal, $window, errorMessages, Errors) {
+    .factory('ErrorService', ['AlertsService', 'errorNames', 'Session', '$log', '$rootScope', '$uibModal', '$window', 'errorMessages', 'Errors', 'UriUtils', 'appNames',
+          function ErrorService(AlertsService, errorNames, Session, $log, $rootScope, $uibModal, $window, errorMessages, Errors, UriUtils, appNames) {
```
Comment (jrchudy) [inline common/errors.js]: `appNames` is not being used in this factory. No need to include it.

File: common/errors.js [hunk 2]
```diff
@@ -214,37 +223,53 @@
...
+               errorPopup(exception.message, exception.status, "Recordset ", exception.errorData.redirectUrl, stackTrace);
+           }
             // we decided to deal with the OR condition later
             else if ( (ERMrest && exception instanceof ERMrest.ForbiddenError) || exception.code == errorNames.forbidden) {
                 errorPopup( exception.message, exception.status ,"Home Page", $window.location.origin);
             }
-            else if ( ERMrest && exception instanceof ERMrest.ConflictError) {
-                errorPopup( exception.message, exception.status ,"Record Page", exception.errorData.redirectUrl, exception.subMessage);
+            else if (ERMrest && exception instanceof ERMrest.ERMrestError ) {
+              if(exception.errorData && exception.errorData.gotoTableDisplayname != 'undeifned' && exception.errorData.gotoTableDisplayname != ''){
```
Comment (jrchudy) [inline common/errors.js]: `undefined` is spelt wrong (`undeifned`).

File: common/errors.js [hunk 3]
```diff
@@ -214,37 +223,53 @@
...
+                gotoLocation = exception.errorData.gotoTableDisplayname;
+              }
+              if(exception.errorData && exception.errorData.redirectUrl != 'undeifned' && exception.errorData.redirectUrl != ''){
```
Comment (jrchudy) [inline common/errors.js]: `undefined` is spelt wrong (`undeifned`).

File: common/errors.js [hunk 4]
```diff
@@ -214,37 +223,53 @@
...
+                errorPopup( exception.message, exception.status, gotoLocation, redirectLink, exception.subMessage);
             }
             else {
                 var errName = exception.status? exception.status:"Terminal Error",
                     errorText = exception.message,
-                    systemAdminMessage = errorMessages.systemAdminMessage;
+                    systemAdminMessage = errorMessages.systemAdminMessage,
+                    redirectLink = $window.location.origin,
+                    pageName = "Home Page",
+                    specialMessage;
```
Comment (jrchudy) [inline common/errors.js]: This variable is not used. Please remove it.

File: common/modal.js [hunk 5]
```diff
@@ -17,20 +17,27 @@
...
         if(vm.params.errorCode == 'Multiple Records Found'){
-          vm.clickActionMessage =  messageMap.recordAvailabilityError.multipleRecords;
+            vm.clickActionMessage =  messageMap.recordAvailabilityError.multipleRecords;
         } else if(vm.params.errorCode == 'Record Not Found'){
-          vm.clickActionMessage = messageMap.recordAvailabilityError.noRecordsFound;
+            vm.clickActionMessage = messageMap.recordAvailabilityError.noRecordsFound;
         } else {
-          vm.clickActionMessage = messageMap.recordAvailabilityError.pageRedirect + vm.params.pageName;
+            vm.clickActionMessage = messageMap.recordAvailabilityError.pageRedirect + vm.params.pageName + '. ';
+            vm.reloadBtn = false;
```
Comment (jrchudy) [inline common/modal.js]: No need to set this to `false` here, it was initialized to `false` up above.

File: common/modal.js [hunk 6]
```diff
@@ -17,20 +17,27 @@
...
+            if(vm.params.appName == 'recordedit'){
+              vm.reloadBtn = true;
+              reloadMessage = ' <p>' + messageMap.terminalError.reloadMessage +' </p>';
+            }
         }
-
+        vm.clickActionMessage += reloadMessage;
```
Comment (jrchudy) [inline common/modal.js]: Is the intention for the `clickActionMessage` to be before an empty `<p>` tag? Do we need the `reloadMessage` if there is no content inside the `<p>` tag?

File: common/errors.js [hunk 7]
```diff
@@ -194,17 +199,21 @@
...
 
                     var win = window.open("", '_blank','width=800,height=600,left=' + x + ',top=' + y);
                     Session.loginInAPopUp(win, reloadCb);
-                }else if(errorCode == errorNames.conflict){
-                   $window.open(redirectLink, '_blank');
                 } else {
-                    $window.location.replace(redirectLink);
+                    if(actionBtnIdentifier == "reload"){
+                        reloadCb();
+                    } else{
```
Comment (jrchudy) [inline common/errors.js]: Add a comment for this else case saying that it handles the case when the modal is closed without a result. This is the case when the `ok` button is selected (the default case).

File: common/modal.js [hunk 8]
```diff
@@ -17,20 +17,27 @@
             $uibModalInstance.dismiss('cancel');
         }
     }])
-    .controller('ErrorModalController', ['$uibModalInstance', 'params', 'messageMap', function ErrorModalController($uibModalInstance, params, messageMap) {
+    .controller('ErrorModalController', ['$uibModalInstance', 'params', 'messageMap', '$window', function ErrorModalController($uibModalInstance, params, messageMap, $window) {
         var vm = this;
         vm.params = params;
         vm.details = false;
         vm.linkText = messageMap.showErrDetails;
+        vm.reloadBtn = false;
```
Comment (jrchudy) [inline common/modal.js]: Rename this to `showReloadBtn`. The name implies it's a button element.

File: common/modal.js [hunk 9]
```diff
@@ -17,20 +17,27 @@
             $uibModalInstance.dismiss('cancel');
         }
     }])
-    .controller('ErrorModalController', ['$uibModalInstance', 'params', 'messageMap', function ErrorModalController($uibModalInstance, params, messageMap) {
+    .controller('ErrorModalController', ['$uibModalInstance', 'params', 'messageMap', '$window', function ErrorModalController($uibModalInstance, params, messageMap, $window) {
         var vm = this;
         vm.params = params;
         vm.details = false;
```
Comment (jrchudy) [inline common/modal.js]: Could you also change this and corresponding calling sites to `showDetails`.

File: common/utils.js [hunk 10]
```diff
@@ -20,6 +20,13 @@
         "*": "/record"
     })
 
+    .constant("appNames", {
```
Comment (jrchudy) [inline common/utils.js]: Is this constant necessary? I don't see it being used anywhere. 

We have this set enumerated already in `ermrestJS` attached to the module component based on their annotation tags.

File: common/utils.js [hunk 11]
```diff
@@ -625,6 +636,13 @@
             return /*@cc_on!@*/false || !!document.documentMode;
         }
 
+        // takes URL path and returns app name
+        function appNamefromUrlPathname(path){
```
Comment (jrchudy) [inline common/utils.js]: Add a comment stating what the expected format for this function is. If I supply a path that includes filters, this won't work.

File: record/record.app.js [hunk 12]
```diff
@@ -121,10 +122,16 @@
                     };
                     // return model;
                     callback(model);
+                    errorData.redirectUrl = $rootScope.reference.unfilteredReference.contextualize.compact.appLink;
+                    errorData.gotoTableDisplayname = $rootScope.reference.displayname.value;
+                    error.errorData = errorData;
```
Comment (jrchudy) [inline record/record.app.js]: Lines 125-127 are unnecessary here. When this error is thrown on line 128, it is caught by the `catch` statement right below it where you set all of these values again.

File: record/record.controller.js [hunk 13]
```diff
@@ -194,17 +195,25 @@
...
+
                 (function (i) {
                     refObj.read(dataModel.pageLimit).then(function (page) {
                         dataModel.page = page;
                         dataModel.rowValues = DataUtils.getRowValuesFromPage(page);
                     }, function (error) {
                         console.log(error);
+                        errorData.redirectUrl = $rootScope.reference.unfilteredReference.contextualize.compact.appLink;
+                        errorData.gotoTableDisplayname = $rootScope.reference.displayname.value;
+                        error.errorData = errorData;
```
Comment (jrchudy) [inline record/record.controller.js]: Same comment as in `record.app.js`

Lines 208-210 are unnecessary here. When this error is thrown on line 211, it is caught by the catch statement right below it where you set all of these values again.

File: common/modal.js (hunk 6 above)
Comment (amitjha21) [reply common/modal.js]: `<p>` tag is added to maintain the space between click action message and buttons. Also maintains consistency in their placement irrespective of reload message

**Comment by amitjha21** [comment]
Will add UI links for error changes.

File: common/modal.js (hunk 9 above)
Comment (amitjha21) [reply common/modal.js]: `showDetails` was already being used. Changed it to `displayDetails`.

**Comment by amitjha21** [comment]
Few links are added as per error scenarios in the main description of the PR.

**Pull request merged**